### DISCUSSION TITLE
help defishing please

### DISCUSSION THREAD

**Discussion by maso-lino** [ImageMagick/ImageMagick#3580] [Help]
Hi there, could anyone help me in defishing a picture?
i know i must use the "-distort barrel" option with some parameters like "0.0 -0.019966 0.0" but i am unable to find the right combination.
i will attach the original picture....
![cam_1](https://user-images.githubusercontent.com/83075533/115877971-4eb4bb80-a448-11eb-99de-f280d7330a79.jpg)

many thanks!

**Comment by snibgo** [comment]
We might assume that (a) the lens is an ideal equidistant fisheye, and (b) the central axis of the lens is at the centre of the image.

Sadly, assumption (b) is false. But with those assumptions, we can use arctan to return the projection to rectilinear. Windows BAT script:
```
set SRC=trento.jpg

FOR /F "usebackq" %%L ^
IN (`%IM%identify ^
  -format "WW4=%%[fx:4*w]\nHH4=%%[fx:4*h]\nW4m1=%%[fx:4*w-1]\nH4m1=%%[fx:4*h-1]" ^
  %SRC%`) ^
DO set %%L

set RAT=6

%IM%convert ^
  -size 1x%HH4% ^
  gradient:black-white ^
  ( +clone ^
    -fx "atan(%RAT%*u)/atan(%RAT%)" ^
  ) ^
  -compose Mathematics -define compose:args=0,0.5,-0.5,0.5 -composite ^
  -scale "%WW4%x%HH4%^!" ^
  pol_map_rel4.png

%IM%convert ^
  %SRC% ^
  -virtual-pixel Black ^
  -define distort:scale=4 ^
  -distort Depolar -1 ^
  pol_map_rel4.png ^
  -compose Displace ^
    -set option:compose:args 0x%H4m1% ^
    -composite ^
  -background Pink -virtual-pixel Background ^
  -define distort:scale=0.25 ^
  -distort Polar -1 ^
  db_depol_pol3.png
```
The result is db_depol_pol3.png. Here is a JPG version of the result:
![d](https://user-images.githubusercontent.com/33812368/115887822-380f6400-a44a-11eb-9f7f-3b7501b2c4c4.jpg)
The number for RAT may not be the best possible.

**Reply by maso-lino** [reply]
many thanks. i will try the bat file.

**Comment by snibgo** [comment]
What camera (and lens) took the photograph? If we know that, we can make a more intelligent guess.

**Reply by maso-lino** [reply]
it's a 5mp chinese camera... the brand is Boavision

https://www.aliexpress.com/item/33027212066.html?spm=a2g0s.9042311.0.0.27424c4d9Mx3RL


Model | HX-HC28391080AS | HX-HC28395MPAS
-- | -- | --
Shape | Bullet
Mount Types | Wall Bracket
Power Supply | DC 12V 1A (Don't included)
IR-CUT | Yes
Performance | Motion detect, video lost, network break,IP conflict, storage device full or error
Network Interface | RJ45 10/100MB Ethernet interface
Wireless | WiFi 802.11 b/g/n
Protocol | Support TCP/IP HTTP/RTSP/FTP/PPPOE/DHCP/DNS/DDNS/NTP/UPnP
Network Transmission | P2P
IP Address | Dynamic IP and static IP address
Alarm Notice | FTP Photo, E-mail alerts
Lens | Megapixel FIsheye Lens(180 Degrees)
image sensor | Hi3518EV200+ 1/3" IMX323 CMOS | Hi3516EV300+ 1/2.8" IMX335 CMOS
Min. Illumination | Color:0.6Lux (F1.2, AGC ON),0 LUX with IR
Video CompressionFormat | H.264 High Profile, support dual stream, AVI format;support 0.1M~6Mpbs bit rate adjustable.
Sensor Size | Progressive Scan
Video Bit rate | 100Kbps~6Mbps
Pixels | 2.0 Megapixel | 5.0 Megapixel
Image Sensor | CMOS
Frame Rate (FPS) | PAL/NTSC:1-15fps
Resolution | 1920*1080 | 2560*1920
Audio Compressionstandard | G.726/G.711
Audio code rate | Rate adaptive
Audio Input | Support, Built-in Microphone
Audio Output | Support Two Way Audio Talk Back(With Speaker)
Infrared Distance | 20m/66ft
Configuration | 3 Array LEDs
Environment | Outdoor, Indoor
Protection level | IP65 / Metal body shell
Operate Temperature | -20~50 degree. humidity less than 90%(non-condensing)

**Comment by snibgo** [comment]
The camera claims to be "Fisheye Lens (180 Degrees)". If this is true, a rectilinear projection would have infinite width. If we assume the field of view from corner to corner is 170 degrees, then we want RAT to be tan(170/2) = 11.43, and the result is:
![x](https://user-images.githubusercontent.com/33812368/115911152-a6155480-a465-11eb-94bc-8677e422945d.jpg)
